src/main/java/com/twcable/jackalope/impl/jcr/NodeImpl.java

```diff
@@ -336,43 +341,122 @@ public boolean hasProperties() throws RepositoryException {
+     */
     @Override
     public NodeType[] getMixinNodeTypes() throws RepositoryException {
-        return new NodeType[0];
+        NodeTypeManager manager = getSession().getWorkspace().getNodeTypeManager();
+        Property property = mixinProperty();
+        Value[] values = property.getValues();
+        if (values == null) return new NodeType[0];
+        NodeType[] types = new NodeType[values.length];
+        for (int i = 0; i < values.length; i++) {
+            types[i] = manager.getNodeType(values[i].getString());
+        }
+        return types;
+    }
+
+
+    private String getJcrNameForQName(String name) throws RepositoryException {
+        return getJcrName(NameFactoryImpl.getInstance().create(name));
+    }
+
+
+    private String getJcrName(Name name) throws RepositoryException {
+        return session.getNamePathResolver().getJCRName(name);
     }
 
 
     @Override
     public boolean isNodeType(String nodeTypeName) throws RepositoryException {
-        return getProperty("jcr:primaryType").getString().equals(nodeTypeName);
+        return primaryTypeProperty().getString().equals(nodeTypeName);
     }
 
 
     @Override
     public void setPrimaryType(String nodeTypeName) throws NoSuchNodeTypeException, VersionException, ConstraintViolationException, LockException, RepositoryException {
-        setProperty("jcr:primaryType", nodeTypeName);
+        setProperty(getJcrNameForQName(Property.JCR_PRIMARY_TYPE), nodeTypeName);
     }
 
 
+    /**
+     * Very simple implementation of Mixin support: Does not check permissions, check for conflicts, etc.
+     */
     @Override
     public void addMixin(String mixinName) throws NoSuchNodeTypeException, VersionException, ConstraintViolationException, LockException, RepositoryException {
-        //Not Implemented
+        Property mixinProperty = mixinProperty();
+        Value[] values = mixinProperty.getValues();
+        final Value[] newValues;
+        if (values == null) {
+            newValues = new Value[1];
+            newValues[0] = new ValueImpl(mixinName);
+        }
+        else {
+            newValues = new Value[values.length + 1];
+            System.arraycopy(values, 0, newValues, 0, values.length);
+            newValues[values.length] = new ValueImpl(mixinName);
+        }
+        mixinProperty.setValue(newValues);
+        session.changeItem(this);
     }
 
 
     @Override
     public void removeMixin(String mixinName) throws NoSuchNodeTypeException, VersionException, ConstraintViolationException, LockException, RepositoryException {
-        //Not Implemented
+        Property mixinProperty = mixinProperty();
+        Value[] values = mixinProperty.getValues();
+        if (values == null) {
+            throw new NoSuchNodeTypeException(mixinName);
+        }
+
+        boolean found = false;
+        Value[] newValues = new Value[values.length - 1];
+        for (int idx = 0, newIdx = 0; idx < values.length; ) {
```

Comment:
I generally avoid style comments, but I will admit that even in Java, I would be include to use a list for newValues and convert back to an array at the end.

I will also note that you don't need idx here so you could probably use a foreach here with newIdx being a counter initialized before the loop begins.

That would also allow you to ditch the found flag, you can just compare newIdx to the size of the original array.